"no entry found for key" panic in helix-view

### Summary

Helix panics (about 1-2 times a day with ~8 hour usage) with the following error:

<details>
<summary>Error</summary>

```
$ RUST_BACKTRACE=full GOPACKAGESDRIVER=$PWD/scripts/gopackagesdriver.sh hx
thread 'main' panicked at 'no entry found for key', helix-view/src/document.rs:1006:10
stack backtrace:
   0:     0x55bd8c72b384 - <unknown>
   1:     0x55bd8bf6965c - <unknown>
   2:     0x55bd8c724d65 - <unknown>
   3:     0x55bd8c72d22e - <unknown>
   4:     0x55bd8c72cf5d - <unknown>
   5:     0x55bd8c5b6b5c - <unknown>
   6:     0x55bd8c72d8bb - <unknown>
   7:     0x55bd8c72d6e4 - <unknown>
   8:     0x55bd8c72b8b4 - <unknown>
   9:     0x55bd8c72d44d - <unknown>
  10:     0x55bd8bee8c33 - <unknown>
  11:     0x55bd8bf67d31 - <unknown>
  12:     0x55bd8bf67cdb - <unknown>
  13:     0x55bd8bee8aa6 - <unknown>
  14:     0x55bd8c51965c - <unknown>
  15:     0x55bd8c217c29 - <unknown>
  16:     0x55bd8c3e7c3f - <unknown>
  17:     0x55bd8c5b3845 - <unknown>
  18:     0x55bd8c5d0fab - <unknown>
  19:     0x55bd8c5a9392 - <unknown>
  20:     0x55bd8c591578 - <unknown>
  21:     0x55bd8c5f27a8 - <unknown>
  22:     0x55bd8c5cdc73 - <unknown>
  23:     0x55bd8c589826 - <unknown>
  24:     0x55bd8c589953 - <unknown>
  25:     0x55bd8c5a908d - <unknown>
  26:     0x55bd8c71f5fb - <unknown>
  27:     0x55bd8c589932 - <unknown>
  28:     0x7f056d9d12d0 - <unknown>
  29:     0x7f056d9d138a - __libc_start_main
  30:     0x55bd8bf1edd5 - <unknown>
  31:                0x0 - <unknown>
```
</details>

I _think_ it happens on save, but not every time I save (I'm saving with `:wa` _very_ frequently because of this issue), but it always catches me off guard, so I can never remember exactly what I was doing when it happened.

I don't think `GOPACKAGESDRIVER` has anything to do with it, it's just what I use to get Go's LSP working with Bazel. I'm pretty confident I've seen this failure even when not running with `GOPACKAGESDRIVER`.

### Reproduction Steps

Unfortunately, I don't know exactly what triggers it. I usually have lots of splits open, and it seems to be either a `:w` or `:wa` command that triggers the failure.

### Helix log

<details><summary>~/.cache/helix/helix.log</summary>

```
2022-08-16T09:11:48.686 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T09:12:15.077 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T09:13:57.611 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T09:46:32.853 helix_view::editor [ERROR] Failed to initialize the LSP for `source.md` { LSP not defined }
2022-08-16T09:46:41.629 helix_view::editor [ERROR] Failed to initialize the LSP for `source.md` { LSP not defined }
2022-08-16T09:48:56.085 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T09:48:59.795 helix_lsp::transport [ERROR] err: <- StreamClosed
2022-08-16T09:49:58.990 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T09:58:05.710 helix_lsp::transport [ERROR] err: <- StreamClosed
2022-08-16T09:58:08.114 helix_lsp::transport [ERROR] err: <- StreamClosed
2022-08-16T09:59:01.288 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T09:59:31.375 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T10:00:07.235 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T10:01:48.057 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T10:03:10.204 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T10:10:03.673 helix_lsp::transport [ERROR] err: <- StreamClosed
2022-08-16T10:10:15.765 helix_view::editor [ERROR] Failed to initialize the LSP for `git.commitmsg` { LSP not defined }
2022-08-16T10:25:57.120 helix_lsp::transport [ERROR] err: <- StreamClosed
2022-08-16T10:46:11.049 helix_lsp::transport [ERROR] <- ServerError(0): 14:21: string literal not terminated (and 2 more errors)
2022-08-16T10:46:11.050 helix_view::document [WARN] LSP formatting failed: protocol error: ServerError(0): 14:21: string literal not terminated (and 2 more errors)
2022-08-16T10:54:14.696 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T10:54:22.729 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T10:55:39.396 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T10:56:59.820 helix_view::editor [ERROR] Failed to initialize the LSP for `source.sql` { LSP not defined }
2022-08-16T10:57:07.583 helix_view::editor [ERROR] Failed to initialize the LSP for `source.sql` { LSP not defined }
2022-08-16T10:59:05.821 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T10:59:40.993 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T10:59:57.991 helix_view::editor [ERROR] Failed to initialize the LSP for `source.sql` { LSP not defined }
2022-08-16T11:04:37.385 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T11:05:12.052 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T11:05:46.337 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T11:08:33.941 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T11:09:06.968 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T11:11:42.715 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T11:12:32.145 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T11:19:10.376 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T11:22:27.091 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T11:23:53.661 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T11:24:57.772 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T11:25:54.567 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T11:26:27.158 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T11:27:15.928 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T11:28:08.760 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T11:29:00.692 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T11:29:32.137 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T11:30:34.538 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T11:31:39.642 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T11:32:53.320 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T11:35:36.503 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T11:41:17.265 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T11:47:09.087 helix_term::application [WARN] lsp position out of bounds - Diagnostic { range: Range { start: Position { line: 142, character: 0 }, end: Position { line: 142, character: 0 } }, severity: Some(Error), code: None, code_description: None, source: Some("syntax"), message: "expected ';', found 'EOF'", related_information: None, tags: None, data: None }
2022-08-16T11:47:09.270 helix_term::application [WARN] lsp position out of bounds - Diagnostic { range: Range { start: Position { line: 143, character: 0 }, end: Position { line: 143, character: 0 } }, severity: Some(Error), code: None, code_description: None, source: Some("syntax"), message: "expected ';', found 'EOF'", related_information: None, tags: None, data: None }
2022-08-16T11:47:09.423 helix_term::application [WARN] lsp position out of bounds - Diagnostic { range: Range { start: Position { line: 144, character: 0 }, end: Position { line: 144, character: 0 } }, severity: Some(Error), code: None, code_description: None, source: Some("syntax"), message: "expected ';', found 'EOF'", related_information: None, tags: None, data: None }
2022-08-16T11:47:15.815 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T11:48:13.727 helix_term::application [WARN] lsp position out of bounds - Diagnostic { range: Range { start: Position { line: 148, character: 0 }, end: Position { line: 148, character: 0 } }, severity: Some(Error), code: None, code_description: None, source: Some("syntax"), message: "expected ';', found 'EOF'", related_information: None, tags: None, data: None }
2022-08-16T11:48:13.898 helix_term::application [WARN] lsp position out of bounds - Diagnostic { range: Range { start: Position { line: 149, character: 0 }, end: Position { line: 149, character: 0 } }, severity: Some(Error), code: None, code_description: None, source: Some("syntax"), message: "expected ';', found 'EOF'", related_information: None, tags: None, data: None }
2022-08-16T11:48:37.571 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T11:49:26.443 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T11:50:01.018 helix_term::application [WARN] lsp position out of bounds - Diagnostic { range: Range { start: Position { line: 154, character: 0 }, end: Position { line: 154, character: 0 } }, severity: Some(Error), code: None, code_description: None, source: Some("syntax"), message: "expected ';', found 'EOF'", related_information: None, tags: None, data: None }
2022-08-16T11:50:01.220 helix_term::application [WARN] lsp position out of bounds - Diagnostic { range: Range { start: Position { line: 155, character: 0 }, end: Position { line: 155, character: 0 } }, severity: Some(Error), code: None, code_description: None, source: Some("syntax"), message: "expected ';', found 'EOF'", related_information: None, tags: None, data: None }
2022-08-16T11:50:01.359 helix_term::application [WARN] lsp position out of bounds - Diagnostic { range: Range { start: Position { line: 156, character: 0 }, end: Position { line: 156, character: 0 } }, severity: Some(Error), code: None, code_description: None, source: Some("syntax"), message: "expected ';', found 'EOF'", related_information: None, tags: None, data: None }
2022-08-16T11:50:39.381 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T11:50:51.460 helix_term::application [WARN] lsp position out of bounds - Diagnostic { range: Range { start: Position { line: 159, character: 0 }, end: Position { line: 159, character: 0 } }, severity: Some(Error), code: None, code_description: None, source: Some("syntax"), message: "expected ';', found 'EOF'", related_information: None, tags: None, data: None }
2022-08-16T11:50:51.629 helix_term::application [WARN] lsp position out of bounds - Diagnostic { range: Range { start: Position { line: 160, character: 0 }, end: Position { line: 160, character: 0 } }, severity: Some(Error), code: None, code_description: None, source: Some("syntax"), message: "expected ';', found 'EOF'", related_information: None, tags: None, data: None }
2022-08-16T11:51:09.160 helix_term::application [WARN] lsp position out of bounds - Diagnostic { range: Range { start: Position { line: 162, character: 0 }, end: Position { line: 162, character: 0 } }, severity: Some(Error), code: None, code_description: None, source: Some("syntax"), message: "expected ';', found 'EOF'", related_information: None, tags: None, data: None }
2022-08-16T11:51:09.309 helix_term::application [WARN] lsp position out of bounds - Diagnostic { range: Range { start: Position { line: 163, character: 0 }, end: Position { line: 163, character: 0 } }, severity: Some(Error), code: None, code_description: None, source: Some("syntax"), message: "expected ';', found 'EOF'", related_information: None, tags: None, data: None }
2022-08-16T11:51:09.460 helix_term::application [WARN] lsp position out of bounds - Diagnostic { range: Range { start: Position { line: 164, character: 0 }, end: Position { line: 164, character: 0 } }, severity: Some(Error), code: None, code_description: None, source: Some("syntax"), message: "expected ';', found 'EOF'", related_information: None, tags: None, data: None }
2022-08-16T11:51:56.141 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T11:52:33.431 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T11:54:42.422 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T11:55:27.757 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T11:56:01.825 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T11:57:10.284 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T11:58:05.591 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"
2022-08-16T11:58:38.891 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <project directory>: open <project directory>/bazel-<project name>/.postgres-data/data: permission denied\n"

```

</details>

Most of these errors are related to `$PROJECT/.postgres-data/data`, which is owned by some other user (it's running in a Docker container). I believe I've seen this failure in other projects that wouldn't have this `.postgres-data` directory, and also when editing non-Go projects (e.g. TypeScript)

### Platform

Linux

### Terminal Emulator

st (4ef0cbd8)

### Helix Version

helix 22.05 (8deaebde) (built from HEAD ~4 days ago)

@dead10ck I'm still getting a panic on the same line (line number has changed in your branch though):

<details>
<summary>stack trace</summary>

```
thread 'main' panicked at 'no entry found for key', helix-view/src/document.rs:1200:10
stack backtrace:
   0:     0x5589b184bed4 - std::backtrace_rs::backtrace::libunwind::trace::h22893a5306c091b4
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x5589b184bed4 - std::backtrace_rs::backtrace::trace_unsynchronized::h29c3bc6f9e91819d
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x5589b184bed4 - std::sys_common::backtrace::_print_fmt::he497d8a0ec903793
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/sys_common/backtrace.rs:66:5
   3:     0x5589b184bed4 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h9c2a9d2774d81873
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/sys_common/backtrace.rs:45:22
   4:     0x5589b106f7dc - core::fmt::write::hba4337c43d992f49
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/fmt/mod.rs:1194:17
   5:     0x5589b18458b5 - std::io::Write::write_fmt::heb73de6e02cfabed
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/io/mod.rs:1655:15
   6:     0x5589b184dd7e - std::sys_common::backtrace::_print::h63c8b24acdd8e8ce
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/sys_common/backtrace.rs:48:5
   7:     0x5589b184dd7e - std::sys_common::backtrace::print::h426700d6240cdcc2
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/sys_common/backtrace.rs:35:9
   8:     0x5589b184dd7e - std::panicking::default_hook::{{closure}}::hc9a76eed0b18f82b
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:295:22
   9:     0x5589b184daad - std::panicking::default_hook::h2e88d02087fae196
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:314:9
  10:     0x5589b16aaddc - helix_term::application::Application::run::{{closure}}::{{closure}}::h2a171355ca53c72c
  11:     0x5589b184e40b - std::panicking::rust_panic_with_hook::habfdcc2e90f9fd4c
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:702:17
  12:     0x5589b184e234 - std::panicking::begin_panic_handler::{{closure}}::he054b2a83a51d2cd
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:588:13
  13:     0x5589b184c404 - std::sys_common::backtrace::__rust_end_short_backtrace::ha48b94ab49b30915
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/sys_common/backtrace.rs:138:18
  14:     0x5589b184df9d - rust_begin_unwind
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:584:5
  15:     0x5589b0feec33 - core::panicking::panic_fmt::h366d3a309ae17c94
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/panicking.rs:143:14
  16:     0x5589b106deb1 - core::panicking::panic_display::hd3e89d99fb893792
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/panicking.rs:72:5
  17:     0x5589b106de5b - core::panicking::panic_str::h98ed104b0c03a108
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/panicking.rs:56:5
  18:     0x5589b0feeaa6 - core::option::expect_failed::h58d487f78ae9d247
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/option.rs:1874:5
  19:     0x5589b163033c - helix_view::document::Document::apply::hdd9730fbde23cca3
  20:     0x5589b1582e05 - core::ops::function::FnOnce::call_once{{vtable.shim}}::he5341b3c91860820
  21:     0x5589b15ad29e - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h236cde45d77243bf
  22:     0x5589b144027e - std::thread::local::LocalKey<T>::with::h440c0bd8e97a5284
  23:     0x5589b154dd68 - futures_executor::local_pool::block_on::hff98863698998444
  24:     0x5589b146a885 - tokio::runtime::enter::exit::hbdc9c7ff73bf1342
  25:     0x5589b13c3ab4 - tokio::runtime::thread_pool::worker::block_in_place::hb38f1bb8e6a60a62
  26:     0x5589b13341db - helix_term::compositor::Context::block_try_flush_writes::hfdbc70b629937314
  27:     0x5589b132d69a - helix_term::commands::typed::write_all_impl::hce3a06ee086ddcf8
  28:     0x5589b1563c79 - helix_term::commands::typed::command_mode::{{closure}}::h162efeea55d3791c
  29:     0x5589b1572ed3 - <helix_term::ui::prompt::Prompt as helix_term::compositor::Component>::handle_event::hd7e09ffd9de34d7b
  30:     0x5589b1334db0 - helix_term::compositor::Compositor::handle_event::h1b818966a61e8398
  31:     0x5589b15c33da - helix_term::application::Application::handle_terminal_events::ha4be59be0fc0081c
  32:     0x5589b16a6b9a - helix_term::application::Application::event_loop_until_idle::{{closure}}::haa95758a1fce8ebe
  33:     0x5589b16c253b - hx::main_impl::{{closure}}::h9515a8a871fef1e9
  34:     0x5589b16df095 - std::thread::local::LocalKey<T>::with::h4a2650524fd0402e
  35:     0x5589b16bb5e8 - tokio::park::thread::CachedParkThread::block_on::h23825ea45ad693b3
  36:     0x5589b1693008 - tokio::runtime::thread_pool::ThreadPool::block_on::h7d48bcf67d5258ff
  37:     0x5589b16f0750 - tokio::runtime::Runtime::block_on::hab8dd3ba388c1c22
  38:     0x5589b16a3666 - hx::main::h8012874220ecbbec
  39:     0x5589b16b2e63 - std::sys_common::backtrace::__rust_begin_short_backtrace::h8b4ea23ef8b6946e
  40:     0x5589b16def7d - std::rt::lang_start::{{closure}}::hc01c98d73647ef30
  41:     0x5589b18401bb - core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once::had4f69b3aefb47a8
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/ops/function.rs:259:13
  42:     0x5589b18401bb - std::panicking::try::do_call::hf2ad5355fcafe775
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:492:40
  43:     0x5589b18401bb - std::panicking::try::h0a63ac363423e61e
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:456:19
  44:     0x5589b18401bb - std::panic::catch_unwind::h18088edcecb8693a
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panic.rs:137:14
  45:     0x5589b18401bb - std::rt::lang_start_internal::{{closure}}::ha7dad166dc711761
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/rt.rs:128:48
  46:     0x5589b18401bb - std::panicking::try::do_call::hda0c61bf3a57d6e6
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:492:40
  47:     0x5589b18401bb - std::panicking::try::hbc940e68560040a9
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:456:19
  48:     0x5589b18401bb - std::panic::catch_unwind::haed0df2aeb3fa368
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panic.rs:137:14
  49:     0x5589b18401bb - std::rt::lang_start_internal::h9c06694362b5b80c
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/rt.rs:128:20
  50:     0x5589b16a3772 - main
  51:     0x7f458d77c2d0 - <unknown>
  52:     0x7f458d77c38a - __libc_start_main
  53:     0x5589b1024f55 - _start
                               at /build/glibc/src/glibc/csu/../sysdeps/x86_64/start.S:115
  54:                0x0 - <unknown>
```

</details>

Nothing interesting in the logs prior to the crash, though I hadn't specified `-vvvv`:

<details>
<summary>helix.log</summary>

```
2022-08-24T10:03:47.565 helix_view::editor [ERROR] Failed to initialize the LSP for `git.commitmsg` { LSP not defined }
2022-08-24T10:06:03.874 helix_view::editor [ERROR] Failed to initialize the LSP for `source.gitattributes` { LSP not defined }
2022-08-24T10:06:57.493 helix_view::editor [ERROR] Failed to initialize the LSP for `git.commitmsg` { LSP not defined }
2022-08-24T10:33:48.962 helix_view::editor [ERROR] Failed to initialize the LSP for `source.bash` { cannot find binary path }
2022-08-24T10:34:26.519 helix_lsp::transport [ERROR] err: <- StreamClosed
2022-08-24T10:34:26.519 helix_lsp::transport [ERROR] err: <- StreamClosed
2022-08-24T10:34:26.521 helix_lsp::transport [ERROR] err: <- StreamClosed
2022-08-24T10:45:05.255 helix_term::application [ERROR] Language Server: Method not found client/registerCapability
2022-08-24T10:45:07.462 helix_term::application [ERROR] Language Server: Method not found window/showMessageRequest
2022-08-24T10:45:07.462 helix_term::application [ERROR] Language Server: Method not found window/showMessageRequest
2022-08-24T10:45:15.017 helix_view::editor [ERROR] Failed to initialize the LSP for `git.commitmsg` { LSP not defined }
2022-08-24T10:50:28.822 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <WORKSPACE>: open <WORKSPACE>/bazel-<PROJECT>/.postgres-data/data: permission denied\n"
2022-08-24T10:53:02.827 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <WORKSPACE>: open <WORKSPACE>/bazel-<PROJECT>/.postgres-data/data: permission denied\n"
2022-08-24T10:53:57.772 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <WORKSPACE>: open <WORKSPACE>/bazel-<PROJECT>/.postgres-data/data: permission denied\n"
2022-08-24T10:57:01.145 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <WORKSPACE>: open <WORKSPACE>/bazel-<PROJECT>/.postgres-data/data: permission denied\n"
2022-08-24T11:04:37.184 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <WORKSPACE>: open <WORKSPACE>/bazel-<PROJECT>/.postgres-data/data: permission denied\n"
2022-08-24T11:42:30.290 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <WORKSPACE>: open <WORKSPACE>/bazel-<PROJECT>/.postgres-data/data: permission denied\n"
2022-08-24T11:43:29.826 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <WORKSPACE>: open <WORKSPACE>/bazel-<PROJECT>/.postgres-data/data: permission denied\n"
2022-08-24T11:44:18.305 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <WORKSPACE>: open <WORKSPACE>/bazel-<PROJECT>/.postgres-data/data: permission denied\n"
2022-08-24T11:44:57.857 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <WORKSPACE>: open <WORKSPACE>/bazel-<PROJECT>/.postgres-data/data: permission denied\n"
2022-08-24T11:45:51.101 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <WORKSPACE>: open <WORKSPACE>/bazel-<PROJECT>/.postgres-data/data: permission denied\n"
2022-08-24T11:46:53.216 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <WORKSPACE>: open <WORKSPACE>/bazel-<PROJECT>/.postgres-data/data: permission denied\n"
2022-08-24T11:49:27.238 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <WORKSPACE>: open <WORKSPACE>/bazel-<PROJECT>/.postgres-data/data: permission denied\n"
2022-08-24T11:50:01.288 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <WORKSPACE>: open <WORKSPACE>/bazel-<PROJECT>/.postgres-data/data: permission denied\n"
2022-08-24T11:50:54.952 helix_lsp::transport [ERROR] err <- "walk.go:91: gopathwalk: scanning directory <WORKSPACE>: open <WORKSPACE>/bazel-<PROJECT>/.postgres-data/data: permission denied\n"
2022-08-24T11:53:58.083 helix_view::editor [ERROR] Failed to initialize the LSP for `source.graphql` { LSP not defined }
```
</details>

--- a/helix-term/src/commands.rs
+++ b/helix-term/src/commands.rs
@@ -2515,19 +2515,20 @@ fn insert_at_line_end(cx: &mut Context) {
 async fn make_format_callback(
     doc_id: DocumentId,
     doc_version: i32,
+    view_id: ViewId,
     format: impl Future<Output = Result<Transaction, FormatterError>> + Send + 'static,
     write: Option<(Option<PathBuf>, bool)>,
 ) -> anyhow::Result<job::Callback> {
     let format = format.await;
 
     let call: job::Callback = Callback::Editor(Box::new(move |editor| {
-        if !editor.documents.contains_key(&doc_id) {
+        if !editor.documents.contains_key(&doc_id) || !editor.tree.contains(view_id) {
             return;
         }
 
         let scrolloff = editor.config().scrolloff;
         let doc = doc_mut!(editor, &doc_id);
-        let view = view_mut!(editor);
+        let view = view_mut!(editor, view_id);
 
         if let Ok(format) = format {
             if doc.version() == doc_version {

--- a/helix-term/src/commands/typed.rs
+++ b/helix-term/src/commands/typed.rs
@@ -270,7 +270,7 @@ fn write_impl(
 ) -> anyhow::Result<()> {
     let editor_auto_fmt = cx.editor.config().auto_format;
     let jobs = &mut cx.jobs;
-    let doc = doc_mut!(cx.editor);
+    let (view, doc) = current!(cx.editor);
     let path = path.map(AsRef::as_ref);
 
     let fmt = if editor_auto_fmt {
@@ -278,6 +278,7 @@ fn write_impl(
             let callback = make_format_callback(
                 doc.id(),
                 doc.version(),
+                view.id,
                 fmt,
                 Some((path.map(Into::into), force)),
             );
@@ -343,9 +344,9 @@ fn format(
         return Ok(());
     }
 
-    let doc = doc!(cx.editor);
+    let (view, doc) = current!(cx.editor);
     if let Some(format) = doc.format() {
-        let callback = make_format_callback(doc.id(), doc.version(), format, None);
+        let callback = make_format_callback(doc.id(), doc.version(), view.id, format, None);
         cx.jobs.callback(callback);
     }
 
@@ -573,12 +574,13 @@ fn write_all_impl(
     let mut errors: Vec<&'static str> = Vec::new();
     let auto_format = cx.editor.config().auto_format;
     let jobs = &mut cx.jobs;
+    let current_view = view!(cx.editor);
 
     // save all documents
     let saves: Vec<_> = cx
         .editor
         .documents
-        .values()
+        .values_mut()
         .filter_map(|doc| {
             if doc.path().is_none() {
                 errors.push("cannot write a buffer without a filename\n");
@@ -589,10 +591,30 @@ fn write_all_impl(
                 return None;
             }
 
+            // Look for a view to apply the formatting change to. If the document
+            // is in the current view, just use that. Otherwise, since we don't
+            // have any other metric available for better selection, just pick
+            // the first view arbitrarily so that we still commit the document
+            // state for undos. If somehow we have a document that has not been
+            // initialized with any view, initialize it with the current view.
+            let target_view = if doc.selections().contains_key(&current_view.id) {
+                current_view.id
+            } else if let Some(view) = doc.selections().keys().next() {
+                *view
+            } else {
+                doc.ensure_view_init(current_view.id);
+                current_view.id
+            };
+
             let fmt = if auto_format {
                 doc.auto_format().map(|fmt| {
-                    let callback =
-                        make_format_callback(doc.id(), doc.version(), fmt, Some((None, force)));
+                    let callback = make_format_callback(
+                        doc.id(),
+                        doc.version(),
+                        target_view,
+                        fmt,
+                        Some((None, force)),
+                    );
                     jobs.add(Job::with_callback(callback).wait_before_exiting());
                 })
             } else {
@@ -602,6 +624,7 @@ fn write_all_impl(
             if fmt.is_none() {
                 return Some(doc.id());
             }
+
             None
         })
         .collect();

